controller/securenaming.go

```diff
@@ -17,31 +17,162 @@
 	serviceIndexer    cache.Indexer
+
+	serviceQueue workqueue.RateLimitingInterface
 }
 
 func NewSecureNamingController(core v1core.CoreV1Interface) *SecureNamingController {
-	si, sc := newServiceIndexerInformer(core)
-	return &SecureNamingController{sc, si}
+	snc := &SecureNamingController{
+		client:       core,
+		mapping:      NewSecureNamingMapping(),
+		serviceQueue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	podHandlers := cache.ResourceEventHandlerFuncs{
+		AddFunc:    snc.addPod,
+		DeleteFunc: snc.deletePod,
+		UpdateFunc: snc.updatePod,
+	}
+	snc.podStore, snc.podController = newPodInformer(core, podHandlers)
+
+	svcHandlers := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			snc.enqueueService(obj.(*v1.Service))
+		},
+		DeleteFunc: func(obj interface{}) {
+			snc.enqueueService(obj.(*v1.Service))
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			snc.enqueueService(cur.(*v1.Service))
+		},
+	}
+	snc.serviceIndexer, snc.serviceController = newServiceIndexerInformer(core, svcHandlers)
+
+	return snc
 }
 
-func newServiceIndexerInformer(core v1core.CoreV1Interface) (cache.Indexer, cache.Controller) {
+func (snc *SecureNamingController) Run(stopCh chan struct{}) {
+	defer snc.serviceQueue.ShutDown()
+
+	go snc.podController.Run(stopCh)
+	go snc.serviceController.Run(stopCh)
+
+	go snc.worker()
+
+	<-stopCh
+}
+
+func (snc *SecureNamingController) worker() {
+	glog.Infof("Starting a worker")
+
+	for snc.processNextService() {
+	}
+}
+
+// TODO: do we need to return a boolean for this function?
+func (snc *SecureNamingController) processNextService() bool {
+	k, quit := snc.serviceQueue.Get()
+	if quit {
+		glog.Infof("No service to be process in the queue; the worker is quitting")
+		return false
+	}
+	defer snc.serviceQueue.Done(k)
+
+	key := k.(string)
+	obj, exist, err := snc.serviceIndexer.GetByKey(key)
+	if err != nil {
+		glog.Warningf("Failed to retrieve a service with key %s", key)
+
+		return true
+	}
+	if !exist {
+		glog.Infof("Service with key %s does not exsit in the Indexer", key)
+
+		// Remove service from the indexer
+		snc.mapping.RemoveService(key)
+		return true
+	}
+
+	svc := obj.(*v1.Service)
+	snc.mapping.AddService(svc.GetName())
+
+	lo := metav1.ListOptions{
+		LabelSelector: labels.Set(svc.Spec.Selector).String(),
+	}
+	pods, err := snc.client.Pods(svc.Namespace).List(lo)
+	if err != nil {
+		glog.Errorf("Failed to get pods for services")
+		return true
+	}
+
+	accounts := []string{}
+	for _, p := range pods.Items {
+		accounts = append(accounts, p.Spec.ServiceAccountName)
+	}
+	snc.mapping.SetServiceAccounts(svc.GetName(), sets.NewString(accounts...))
+
+	return true
+}
+
+func (snc *SecureNamingController) addPod(obj interface{}) {
+	p := obj.(*v1.Pod)
+	snc.enqueueServices(snc.getPodServices(p))
+}
+
+func (snc *SecureNamingController) deletePod(obj interface{}) {
+	p := obj.(*v1.Pod)
+	snc.enqueueServices(snc.getPodServices(p))
```

Comment:
As discussed offline, suggest to combine addPod and deletePod since they have identical content. Maybe rename it to some generic name.